### PR TITLE
mapAccum

### PR DESCRIPTION
**Pull request by jcmartin**
This is a part of Issue #89. Ideally, all the functions in container should be provided in unordered-containers. mapAccum has been added with test functions. notMember has also been added since it was trivial to add.

### CODE REVIEW

File: Data/HashMap/Array.hs [hunk 1]
```diff
@@ -399,6 +407,20 @@ map f = \ ary ->
              go ary mary (i+1) n
 {-# INLINE map #-}
 
+mapAccum :: (a -> b -> (a,c)) -> a -> Array b -> (a,Array c)
```
Comment (tibbe) [inline Data/HashMap/Array.hs]: Please do.

File: Data/HashMap/Base.hs [hunk 2]
```diff
@@ -576,6 +584,7 @@ adjust f k0 m0 = go h0 k0 0 m0
         | otherwise = t
 {-# INLINABLE adjust #-}
 
+
```
Comment (tibbe) [inline Data/HashMap/Base.hs]: Please get rid of unnecessary whitespace changes, here and elsewhere.

File: Data/HashMap/Base.hs [hunk 3]
```diff
@@ -745,6 +753,29 @@ map :: (v1 -> v2) -> HashMap k v1 -> HashMap k v2
 map f = mapWithKey (const f)
 {-# INLINE map #-}
 
+-- | /O(n)/. The function 'mapAccum' threads an accumulating
+-- argument through the map in some unspecified order of keys.
+mapAccum :: (a -> v1 -> (a,v2)) -> a -> HashMap k v1 -> (a,HashMap k v2)
+mapAccum f = mapAccumWithKey (\a' _ x' -> f a' x')
+
+-- | /O(n)/. The function 'mapAccumWithKey' threads an accumulating
```
Comment (tibbe) [inline Data/HashMap/Base.hs]: You need a lazy version and a strict version. The lazy one shouldn't evaluate the value thunk before putting it in the map.

There's also the issue of the strictness of the accumulator. We need to document it's strictness (i.e. is the accumulator treated as in `foldl` or in `foldl'`.)

**Comment by tibbe** [comment]
Please rebase and squash the commits so we don't make the history too messy (which would make e.g. `git bisect` work less well.)

**Comment by treeowl** [comment]
@tibbe, `findWithDefault` has actually been added here following a brief discussion on the libraries list started by @m-renaud. I hadn't read this discussion before ultimately approving that. Would you like me to revert before release? I'm not a big fan of the function myself; I'd rather have something like

```haskell
fooLook :: r -> (v -> r) -> k -> HashMap k v -> r
```

**Comment by tibbe** [comment]
I don't care much at this point to be honest. Up to you.

**Comment by sjakobi** [comment]
Here's my current perspective:

1. I'm not yet convinced that replicating the full `Data.Map` API in `Data.HashMap` is a good goal. If there's demand for that, I think it should be discussed with the new maintainers (@emilypi and me in addition to @treeowl) first (in a new issue).

2. `mapAccum[WithKey]` doesn't strike me as a particularly in-demand addition. Can't `traverseWithKey` be used for the same purpose?

   >The function 'mapAccum' threads an accumulating argument through the map in some unspecified order of keys.

   When the order of keys is unspecified, I think it will be only useful for commutative operations, and therefore quite limited…

3. `notMember` doesn't pass the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold) for me.

**Comment by sjakobi** [comment]
Sorry, that was a bit harsh.

I very much do value contributions, but in this case they should have been coordinated a bit better first. In particular, when we're adopting functionality from `containers`, we need to consider whether it is still as useful in a context where the order of keys is not ascending and unstable.

My vote is for closing this PR.

**Comment by jcmartin** [comment]
Sorry, I forgot about this PR, and I will at least give my perspectives from what I recall.

> Here's my current perspective on this PR:
> 
> 1. I'm not yet convinced that replicating the full `Data.Map` API in `Data.HashMap` is a good goal. If there's demand for that, I think it should be discussed with the new maintainers (@emilypi and me in addition to @treeowl) first (in a new issue).

There are occasions where one needs to a _drop in replacement_ from Map to HashMap or vice versa. It was such an occasion that prompted me to do this PR to begin with. In addition, it removes some _cognitive overhead_ if one get away with using a similar, if not the same, API.

> 2. `mapAccum[WithKey]` doesn't strike me as a particularly in-demand addition. Can't `traverseWithKey` be used for the same purpose?
>    > The function 'mapAccum' threads an accumulating argument through the map in some unspecified order of keys.
>    
>    
>    When the order of keys is unspecified, I think it will be only useful for commutative operations, and therefore quite limited…

This is true that the `traverseWithKey` could be used for the same purpose. However, it conflicts with the first point.

> 3. `notMember` doesn't pass the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold) for me.

That is a fair assessment. However, I would vote that `notMember` be also removed from containers as well.

**Comment by tibbe** [comment]
I think the main reason I was reluctant to add new functions for completeness sake (compared to `Data.Map`) is that it makes it harder to improve the library internally. Concretely, any change to the internal data structure typically requires changing the implementation of all functions. More functions means fewer internal improvements due to that work load increase.

On a more philosophical note I feel it's a shame we can't good performance and expressibility in Haskell with only a few operator. For example, in C++ `(const) T& operator[]` can be used for all single element modifications and access, while we have a whole bunch of functions for each use case. We once tried to use zippers to achieve the same in Haskell but performance was dreadful.

**Comment by sjakobi** [comment]
@jcmartin 

> There are occasions where one needs to do a _drop in replacement_ from Map to HashMap or vice versa.

Yeah. I've wanted to do the same before. The tricky bit is that many Map operations work with the ascending key order.

I suspect that the easiest way to migrate Map code to HashMap and vice versa is to use something like [`mono-traversable`'s `IsMap` class](https://hackage.haskell.org/package/mono-traversable-1.0.15.1/docs/Data-Containers.html#t:IsMap).

> However, I would vote that `notMember` be also removed from containers as well.

That's not unfair. The issue is that once a function is exported, it can be costly to get rid of it.

@tibbe 

> I think the main reason I was reluctant to add new functions for completeness sake (compared to `Data.Map`) is that it makes it harder to improve the library internally. Concretely, any change to the internal data structure typically requires changing the implementation of all functions. More functions means fewer internal improvements due to that work load increase.

Yes, that's another important consideration. With the new maintainership it seems more likely that the internals will change in a significant way. Maybe we should hold off on the `containers`-compatibility issue for a few months.

**Comment by sjakobi** [comment]
Closing for the reasons mentioned in https://github.com/haskell-unordered-containers/unordered-containers/pull/91#issuecomment-638779035.

I would welcome a documentation PR that adds a `mapAccum`ish example to the haddocks for `traverseWithKey` though.
